add support for datasketches js udafs using wasm libraries

nikunjbhartia opened the pull request:
Adding support for 3 datasketches sketch types : 
1) Theta sketch 
2) KLL sketch 
3) Tuple sketch 

Total functions: 
- 9 Aggregate functions ( JS UDAFs) 
- 7 scalar functions  ( JS UDF) 

Solution approach: Datasketch cpp libraries are compiled to WASM libraries using emscripten toolchain and loaded in BQ JS UDAFs. using docker.io emscripten image in build file

Note: currently unit tests are executed using `dataform test` which do no support UDAFs or constant arguments. 
Hence, added test cases for 7 scalar functions only 

There is also a datasketches-cpp directory which could be added as a submodule, committed all the files for now, lets discuss during review.

danieldeleo commented:
I've submitted https://github.com/GoogleCloudPlatform/bigquery-utils/pull/425 to enable unit testing for UDAFs. Once it's merged, please pull the changes and add the remaining 9 test cases for the UDAFs

nikunjbhartia commented:
Thanks for adding support for UDAF unit testing. I have added unit test for 1 UDAF which did not have any const arg in its signature. Rest of the 8 UDAFs have const args and hence can't be added yet. 

Current framework uses positional args which would fail for functions having const args. we need to add additional support for passing constant args  which are defined as "NOT AGGREGATE" in UDAF function signature.

danieldeleo commented:
@nikunjbhartia you should be good to go for the rest of the 8 UDAFs now that https://github.com/GoogleCloudPlatform/bigquery-utils/pull/426 is merged

nikunjbhartia commented:
Thanks @danieldeleo for the quick fix. Made following updates: 
- Added unit tests for rest of the 8 UDAFs with const args
- Removed datasketches-cpp folder 
- Added a build task for cloning the above repo

danieldeleo commented on udfs/cloudbuild.yaml:
```diff
@@ -53,6 +53,27 @@ steps:
   args:
     - run-script
     - build-all-libs
+  ###############################################################################
+  # Clone datasketches-cpp github repo as dependency for building WASM libraries
+  ###############################################################################
+- name: gcr.io/cloud-builders/git
+  id: clone_datasketches_cpp_github
+  dir: datasketches
+  args: [ 'clone', 'https://github.com/apache/datasketches-cpp.git' ]
+  #########################################################################
+  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
+  #########################################################################
+- name: docker.io/emscripten/emsdk
+  id: build_wasm_libs
+  dir: datasketches
+  entrypoint: bash
+  args:
+  - '-c'
+  - |
+    for dir in kll-sketch theta-sketch tuple-sketch; do
+      cd $dir && make clean && make all
+      cd ..
+    done
```
```suggestion
#########################################################################
  # Clone datasketches-cpp github repo as dependency for building WASM libraries
  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
  #########################################################################
- name: docker.io/emscripten/emsdk
  id: build_wasm_libs
  dir: datasketches
  entrypoint: bash
  args:
  - '-c'
  - |
    git clone https://github.com/apache/datasketches-cpp.git
    for dir in kll-sketch theta-sketch tuple-sketch; do
      cd $dir && make clean && make all
      cd ..
    done
```

danieldeleo commented on udfs/cloudbuild.yaml:
```diff
@@ -53,6 +53,22 @@ steps:
   args:
     - run-script
     - build-all-libs
+#########################################################################
+  # Clone datasketches-cpp github repo as dependency for building WASM libraries
+  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
+  #########################################################################
```
```suggestion
 
 #########################################################################
  # Clone datasketches-cpp github repo as dependency for building WASM libraries
  # Build (via emscripten) all WASM libraries for BigQuery Sketch UD(A)Fs
  #########################################################################
```

danieldeleo commented on udfs/community/kll_sketch_float64.sqlx:
```diff
@@ -0,0 +1,170 @@
+config { hasOutput: true }
+/*
+ * Copyright 2021 Google LLC
```
```suggestion
 * Copyright 2024 Google LLC
```

danieldeleo replied:
Update all other new files to be 2024 as well

nikunjbhartia replied:
Updated

danieldeleo commented on udfs/datasketches/README.md:
```diff
@@ -0,0 +1,291 @@
+select
+    group_key,
+    count(*) as total_count,
+    `$BQ_PROJECT.$BQ_DATASET`.theta_sketch_int64(x,14) as theta_sketch
```
Now that we're publishing all the UDFs and UDAFs in the bqutil.fn dataset, let's save the user some manual work and just have all the examples in this README reference the public UDFs

nikunjbhartia replied:
Updated

danieldeleo commented on udfs/datasketches/kll-sketch/crypto.js:
```diff
@@ -0,0 +1 @@
+var crypto = { getRandomValues: (array) => { for (var i = 0; i < array.length; i++) array[i] = (Math.random()*256)|0 } };
```
Please add license header

nikunjbhartia replied:
Updated

danieldeleo commented on udfs/datasketches/kll-sketch/kll_sketch.cpp:
```diff
@@ -0,0 +1,92 @@
+#include <stdio.h>
```
Please add license header here and to all other source files

nikunjbhartia replied:
Updated

danieldeleo commented:
/gcbrun
